### ISSUE TITLE
Implement infrastructure for creating empty WordPress draft post based on Idea Hub idea

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#3269]
REST and data store infrastructure for creating a draft post based on an Idea Hub idea should be implemented, based on the foundation established in #3154 and #3157.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The `POST:create-idea-draft-post` REST data point should be implemented:
    * It should expect a parameter `idea` which should be an associative array (object in JS) representing an idea, including the three expected fields `name` (string), `text` (string) and `topics` (associative array) (see also #3155).
    * It should create a new empty WordPress draft post.
        * The [`wp_insert_post_empty_content` filter](https://developer.wordpress.org/reference/hooks/wp_insert_post_empty_content/) needs to be temporarily forced to `__return_false` in order to allow for an empty post to be created.
    * It should then rely on the `set_post_idea` method (see #3157) to associate the new draft post with the idea.
    * The endpoint should return the passed idea associative array/object, expanded with the following fields:
        * `postID` (integer, ID of the new draft post)
        * `postEditURL` (string, URL to the editor for the new draft post)
        * these two fields should be at the same level as the three existing idea fields
* The `modules/idea-hub` JS store should get a new action `createIdeaDraftPost( idea )`:
    * It should rely on the `POST:create-idea-draft-post` REST data point above, passing through `idea` as the `idea` parameter, and returning the response as is (or a potential error).

## Implementation Brief

**PHP**
Using `includes/Modules/Idea_Hub.php` (you can reference the same function name in other modules to see how they have implemented things)
Update `POST:create-idea-draft-post` in `create_data_request`
Check for `$data['idea']` containing the following three properties, `name` (string), `text` (string), `topics`( assoc array), see `GET:new-ideas` for an example of the format for each of these params. If these properties are missing, raise appropriate `WP_Error`s - example: https://github.com/google/site-kit-wp/blob/15bf4bde55b5b6cec126a207c367dd33aefe5d3b/includes/Modules/Analytics.php#L369
If there are no errors:
Create a new draft empty post (the `wp_insert_post_empty_content` filter needs to be temporarily forced to `__return_false` in order to allow for an empty post to be created.).
Call set_post_idea() ( see https://github.com/google/site-kit-wp/issues/3157 ) with the idea and post id.
Return the post id.

In `parse_data_response`, for the same `POST:create-idea-draft-post`:
Return the three fields of the `idea`: `name`, `text`, `topics`, also `postID` (new post we just created), `postEditURL` (url to editor for the post, see https://developer.wordpress.org/reference/functions/get_edit_post_link/ )

**JS**
Create a new fetch store for the `POST:create-idea-draft-post` endpoint `/assets/js/modules/idea-hub/datastore/draft-ideas.js` (#3270 will also use this file)
Use createIdeaDraftPost base name.
Issue a POST request to the modules/idea-hub/data/create-idea-draft-post endpoint with `idea` param.
Validate the idea in `validateParams`, `name`, `text` and `topics`
`reducerCallback` should return the `ideaDraftPost` along with the state
Create a new `baseActions` object `createIdeaDraftPost( idea )` which makes a call to the api endpoint
`const { response, error } = yield fetchCreateIdeaDraftPostStore.actions.fetchCreateIdeaDraftPost( idea );`
Import the create-idea-draft-post store in the datastore/index.js of the idea-hub module.

### Test Coverage

* Add js tests for the new data store.

### Visual Regression Changes

* N/A

## QA Brief

* Enable the idea hub module via tester plugin, open the console on `/wp-admin/admin.php?page=googlesitekit-dashboard`,
```js
googlesitekit.data.stores["modules/idea-hub"].actions.createIdeaDraftPost({
    "name": "ideas/7612031899179595408",
    "text": "How to speed up your WordPress site",
    "topics": [
      {
        "mid": "/m/09kqc",
        "display_name": "Websites"
      }
    ]
  })
```

Look in your network tab, xhr for your request, the response should look something like this:
```
{
    "idea": {
        "name": "ideas/7612031899179595408", 
        "text": "How to speed up your WordPress site", 
        "topics": [
            {
                "mid": "/m/09kqc", 
                "display_name": "Websites"
            }
        ], 
        "postID": 45, 
        "postEditURL": "http://sitekit.10uplabs.com/wp-admin/post.php?post=45&action=edit"
    }
}
```

Take note of that postID and using wp-cli ensure you have the post meta set correctly `$ wp post meta list 45`

## Changelog entry

* Add Idea Hub draft post support.

**Comment by eugene-manuilov** [comment]
Thanks, @ivankruchkoff. In general, it looks good, but we need to make a few adjustments:

> Return the three fields of the idea: name, text, topics, also postID (new post we just created), postEditURL (url to editor for the post, see https://developer.wordpress.org/reference/functions/get_edit_post_link/ )

We need to do it in the `parse_data_response` method. It will let us reuse it by one of the endpoints from #3270. The handler in the  `create_data_request` method should create a new draft post, assign an idea to it, and return a post ID.

> Create a new fetch store for the POST:create-idea-draft-post endpoint /assets/js/modules/idea-hub/datastore/create-idea-draft-post.js

Let's call the file to be `draft-ideas.js`. #3270 will use the same file for draft post related selector.

**Comment by ivankruchkoff** [comment]
Updated @eugene-manuilov. The response may be the same, but I think the way it's built will differ between #3270 and this `parse_data_response` as our request already has all of the params to build it out, and in that ticket there's a WP_Query to leverage for getting that data.

**Comment by eugene-manuilov** [comment]
QA :warning: Although this feature works as expected, there are a few things in the code that need to be fixed:

1. A new post should be created inside of the closure returned as a result for the `POST:create-idea-draft-post` case. https://github.com/google/site-kit-wp/blob/7abc4db5a51c7e83daccd56bdc09220fc3f0637e/includes/Modules/Idea_Hub.php#L251-L263 This code ^ should be moved into here: https://github.com/google/site-kit-wp/blob/7abc4db5a51c7e83daccd56bdc09220fc3f0637e/includes/Modules/Idea_Hub.php#L265-L267
1. The response parser handler shouldn't use idea data stored in the `$data` variable. It should use idea information returned from the `get_post_idea` method. https://github.com/google/site-kit-wp/blob/7abc4db5a51c7e83daccd56bdc09220fc3f0637e/includes/Modules/Idea_Hub.php#L436-L446 Plus, I think we can re-use response parser for the `GET:draft-post-ideas` endpoint (I know it didn't exist at that time yet) since it does return the same data:
   ```php
   case 'POST:create-idea-draft-post':
   case 'GET:draft-post-ideas':
       return array_filter(
           ...
       );
   ```
1. We also need to remove the temporary filter right after creating a new draft post. https://github.com/google/site-kit-wp/blob/7abc4db5a51c7e83daccd56bdc09220fc3f0637e/includes/Modules/Idea_Hub.php#L251-L254
1. I know this is mentioned in the IB, but we don't need to save created draft post in the idea-hub datastore, AC doesn't mention it. https://github.com/google/site-kit-wp/blob/7abc4db5a51c7e83daccd56bdc09220fc3f0637e/assets/js/modules/idea-hub/datastore/draft-ideas.js#L37-L42

**Comment by ivankruchkoff** [comment]
1. Moved to the closure in 
https://github.com/google/site-kit-wp/pull/3391/commits/89be79629ad8ade813bdb345da15f8fdf07ffadf

2. Updated to use the idea from meta in
https://github.com/google/site-kit-wp/pull/3391/commits/8e434e7ab4e2fe38dfca29e54bcb8ace7faf2094

3. I don't think there's a reason to remove this filter, as it's only added in a very specific part of the code, and won't be accidentally reached. Can you think of a scenario where we'd reach the codepath that adds the filter `POST:create-idea-draft-post` but then need it removed?

4. Updated to remove the state changes
https://github.com/google/site-kit-wp/pull/3391/commits/296b31ab5b6609986cd06f7ac131295269626d51

**Comment by eugene-manuilov** [comment]
> 3. I don't think there's a reason to remove this filter, as it's only added in a very specific part of the code, and won't be accidentally reached. Can you think of a scenario where we'd reach the codepath that adds the filter `POST:create-idea-draft-post` but then need it removed?

I don't have a real example, but there might be a plugin that listens to different REST API filters and might be affected by our hook that we decided to leave. It's just a good habit to remove temporary hooks once they become unnecessary.

**Comment by ivankruchkoff** [comment]
Back over to you @eugene-manuilov 

For the console:

```js
googlesitekit.data.stores["modules/idea-hub"].store.getState().draftPostIdeas
```

```js
googlesitekit.data.stores["modules/idea-hub"].actions.createIdeaDraftPost({
    "name": "FOO/7612031899179595408",
    "text": "How to FAST up your WordPress site",
    "topics": [
      {
        "mid": "/m/09kqc",
        "display_name": "AHA"
      }
    ]
  })
``` 

```js
googlesitekit.data.stores["modules/idea-hub"].store.getState().draftPostIdeas
```

**Comment by asvinb** [comment]
Response:
![image](https://user-images.githubusercontent.com/31818/119649108-4e3a8680-be33-11eb-844d-e52f270394ba.png)

Meta all set:
![image](https://user-images.githubusercontent.com/31818/119649269-7f1abb80-be33-11eb-8cf4-02a1dd32d7ce.png)

### QA: ✅

**Comment by felixarntz** [comment]
This was already part of the 1.34.0 release, but it's not user-facing so let's reopen this issue to fix that point I just mentioned above. @ivankruchkoff Unless there is something I missed, can you please open a follow-up PR against `develop` for changing the return format?

**Comment by eugene-manuilov** [comment]
> I just noticed now (via [#3357 (comment)](https://github.com/google/site-kit-wp/issues/3357#issuecomment-856660512)) that the endpoint here is not exactly returning what is mentioned in the ACs:
> 
> * It currently returns an array with the idea associative array (object in JS) for which the post was generated.
> * It should instead return just the idea associative array (object in JS) for which the post was generated.
> 
> There will never be multiple posts created - was there a particular reason that I missed why this was diverged from the ACs?

@felixarntz I think it's my fault :man_facepalming:. I asked Ivan to re-use the `'GET:draft-post-ideas'` handler in the `Idea_Hub::parse_data_response` method (which has the similar functionality but intended for multiple posts) to avoid duplicate code: https://github.com/google/site-kit-wp/blob/d2e73634ec8a3a1a1354530cc67fdc4902aa6655/includes/Modules/Idea_Hub.php#L406-L421

**Comment by felixarntz** [comment]
@eugene-manuilov Thanks, no worries. Let's just fix it to unblock #3357 :)

**Comment by eugene-manuilov** [comment]
Ok, i'll create a PR for it.

**Comment by eugene-manuilov** [comment]
@felixarntz do you mind reviewing #3515?

**Comment by johnPhillips** [comment]
QA ✅

Making the XHR request from the console as described in the QAB worked:

![Screenshot 2021-06-10 at 09 43 49](https://user-images.githubusercontent.com/2470911/121494557-c32fc380-c9d0-11eb-9a45-0e60369a7aed.png)

Querying WP CLI produced the following metadata:
```
+---------+---------------------------------+----------------------------------+
| post_id | meta_key                        | meta_value                       |
+---------+---------------------------------+----------------------------------+
| 49      | googlesitekitpersistent_idea_na | ideas/7612031899179595408        |
|         | me                              |                                  |
| 49      | googlesitekitpersistent_idea_te | How to speed up your WordPress s |
|         | xt                              | ite                              |
| 49      | googlesitekitpersistent_idea_to | a:1:{i:0;a:2:{s:3:"mid";s:8:"/m/ |
|         | pics                            | 09kqc";s:12:"display_name";s:8:" |
|         |                                 | Websites";}}                     |
+---------+---------------------------------+----------------------------------+
```

In addition, the endpoint now returns the draft post object to the `reducerCallback` rather than the array mentioned in https://github.com/google/site-kit-wp/issues/3357#issuecomment-856660512:
```js
// assets/js/modules/idea-hub/datastore/draft-ideas.js: 37
reducerCallback: ( state, ideaDraftPost ) => {
+	console.log( { ideaDraftPost } );
	return {
		...state,
		draftPostIdeas: [ ...( state.draftPostIdeas || [] ), ideaDraftPost ],
	};
},
/* Outputs the following:
{
    "ideaDraftPost": {
        "postID": 45,
        "postEditURL": "http://sitekit.10uplabs.com/wp-admin/post.php?post=45&amp;action=edit",
        "name": "ideas/7612031899179595408",
        "text": "How to speed up your WordPress site",
        "topics": [
            {
                "mid": "/m/09kqc",
                "display_name": "Websites"
            }
        ]
    }
}
*/
```
